### PR TITLE
Improves bad proc arg handling

### LINKED ISSUE #265: Can't pass `null` as a proc arg
`/obj/structure/spacevine/temperature_expose(null, temp, volume)`

### LINKED ISSUE #475: Modified types don't support a space before the {
I'm too tired to fix it rn but I don't want to forget about it.

### PR DESCRIPTION
**Pull request by ike709**
Fixes #265 
Fixes #475 

The error now matches BYOND (with the addition of the token string for the bad arg). Unlike BYOND it also errors for http://www.byond.com/forum/post/2736128

Error recovery should be robust enough to not need to skip.

Also slipped in a fix for #475 and fixed error recovery for `'else if'` switch cases.

Coded this at 4am so it's entirely possible that it could be done more efficiently, but it works and it's an error branch.

### CODE REVIEW

File: DMCompiler/Compiler/DM/DMParser.cs [hunk 1]
```diff
@@ -1405,22 +1406,64 @@ public DMASTProcStatementLabel Label(DMASTIdentifier expression)
...
+                    parameters.AddRange(DefinitionParameters());
+                }
+
                 while (Check(TokenType.DM_Comma)) {
                     BracketWhitespace();
-
+                    if (Current().Location.Line == 83 && Current().Location.Column == 71)
+                    {
+                        Console.WriteLine("e");
+                    }
```
Comment (ike709) [inline DMCompiler/Compiler/DM/DMParser.cs]:
```suggestion
```

File: DMCompiler/Compiler/DM/DMParser.cs [hunk 2]
```diff
@@ -1405,22 +1406,60 @@ public DMASTProcStatementLabel Label(DMASTIdentifier expression)
             List<DMASTDefinitionParameter> parameters = new();
             DMASTDefinitionParameter parameter = DefinitionParameter();
 
-            if (parameter != null || Check(TokenType.DM_IndeterminateArgs)) {
+            if (parameter == null && !Check(TokenType.DM_IndeterminateArgs) && Current().Type != TokenType.DM_RightParenthesis && Current().Type != TokenType.EndOfFile)
+            {
+                // Breaking change - BYOND doesn't specify the arg
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DM/DMParser.cs:1411]: Is this really a "breaking" change?

File: DMCompiler/Compiler/DM/DMParser.cs [hunk 3]
```diff
@@ -1405,22 +1406,60 @@ public DMASTProcStatementLabel Label(DMASTIdentifier expression)
...
+                Error($"error: bad argument definition '{Current().PrintableText}'", false);
+                Recover();
+
+            }
+            else {
                 if (parameter != null) parameters.Add(parameter);
 
+                BracketWhitespace();
+                if (Current().Type != TokenType.DM_RightParenthesis && Current().Type != TokenType.DM_Comma && !Check(TokenType.DM_IndeterminateArgs))
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DM/DMParser.cs:1420]: `DefinitionParameters()` is intended to only parse the vars (`var/a, var/b, var/c`) and not the parentheses. This check (and others) should be moved to `ProcCall()`.

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: DMCompiler/Compiler/DM/DMParser.cs (hunk 2 above)
Comment (ike709) [inline DMCompiler/Compiler/DM/DMParser.cs:1411]:
```suggestion
                // BYOND doesn't specify the arg
```

**Comment by Hinaichigo** [comment]
Will this also fix the error `Duplicate argument "null"` when `null` is passed as an arg multiple times to a proc?

**Comment by wixoaGit** [comment]
> Will this also fix the error `Duplicate argument "null"` when `null` is passed as an arg multiple times to a proc?

Are you talking about procs defined like this?

`/proc/foo(null, null)`

If so, this won't fix that. It was an intentional decision to make 2 arguments with the same name a compile error. These are also not null values being passed to the proc, this is creating a var named `null` that shadows (and replaces) any uses of `null` which is likely unintentional.

**Comment by Hinaichigo** [comment]
That's what I meant, yeah.
So allowing things like:
```
/obj/item/weapon/sword/executioner/afterattack(null, mob/living/user as mob|obj, null, null, null)
	if(complete)
		user.delayNextAttack(30) //thrice the regular attack delay
```
is another silent error with BYOND?

**Comment by wixoaGit** [comment]
> That's what I meant, yeah. So allowing things like:
> 
> ```
> /obj/item/weapon/sword/executioner/afterattack(null, mob/living/user as mob|obj, null, null, null)
> 	if(complete)
> 		user.delayNextAttack(30) //thrice the regular attack delay
> ```
> 
> is another silent error with BYOND?

Yes, multiple parameters with the same name doesn't raise any errors in BYOND. Naming a parameter `null` is actually another issue.

**Comment by ike709** [comment]
Correct

**Comment by Hinaichigo** [comment]
So it's an issue masked by another issue. That makes sense.

File: DMCompiler/Compiler/DM/DMParser.cs [hunk 4]
```diff
@@ -166,7 +168,24 @@ public DMASTStatement Statement(bool requireDelimiter = true)
                     if (Check(TokenType.DM_LeftParenthesis)) {
                         DMCompiler.VerbosePrint($"Parsing proc {_currentPath}()");
                         BracketWhitespace();
-                        DMASTDefinitionParameter[] parameters = DefinitionParameters();
+                        var parameters = DefinitionParameters();
+
+                        if (Current().Type != TokenType.DM_RightParenthesis && Current().Type != TokenType.DM_Comma &&
+                            !Check(TokenType.DM_IndeterminateArgs))
+                        {
+                            Error($"error: {parameters.Last().Name}: missing comma ',' or right-paren ')'", false);
```
Comment (wixoaGit) [inline DMCompiler/Compiler/DM/DMParser.cs:176]: `parameters.Last()` throws an exception if there aren't any parameters.

For example: `proc/t(`

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**